pkg/endpoint/endpoint.go

```diff
@@ -1759,13 +1762,22 @@ type MetadataResolverCB func(ns, podName string) (pod *slim_corev1.Pod, _ []slim
 func (e *Endpoint) RunMetadataResolver(restoredEndpoint, blocking bool, baseLabels labels.Labels, bwm bandwidth.Manager, resolveMetadata MetadataResolverCB) (regenTriggered bool) {
 	done := make(chan bool)
 	regenTriggeredCh := make(chan bool)
-	controllerName := resolveLabels + "-" + e.GetK8sNamespaceAndPodName()
+
+	// make the indexer name
+	e.unconditionalLock()
+	idx := e.resolverIdx
+	e.resolverIdx++
+	e.unlock()
+	controllerName := fmt.Sprintf("%s-%d-%s", resolveLabels, idx, e.GetK8sNamespaceAndPodName())
```

Comment:
In terms of controller-lifetime, this looks good to me now. But this raises the question:

Let's assume `resolve-endpoint-1-xwing` is called with `baseLabels={foo:bar1}`, but it fails a few times (e.g. because api-server is overloaded and resolveMetadata returns errors) and is in a back-off period for a couple of seconds.

Meanwhile, api-server recovers, `resolve-endpoint-2-xwing` is called with `baseLabels={foo:bar2}`, and immediately succeeds. Then, the  `resolve-endpoint-1-xwing` is woken up again, also suceeds, but now overwrites the labels with the outdated `baseLabels={foo:bar1}`.

Is that a problem? Should we somehow track the resolverIndex inside `DoFunc` and give up if we see that we got replaced by a different controller?

Reply:
I can't help but feel like `pkg/controller` should just have an option to "Remove controller once complete" so that we don't have this code in `pkg/endpoint` trying to make sure that multiple separate calls into this function don't go and delete other references of the controller with the same name. Creating multiple controllers and tracking these indexes and so on just seems more complicated than we need to ensure that there's only one copy of a controller at once and that it is removed dynamically once complete.

Incidentally, now that this controller doesn't have a `RunInterval`, do we even need to remove it after completion at all?

Reply:
Nice, a solution where we basically just delete code :100: